src/boot/efi/efi-string.c

```diff
@@ -878,16 +878,30 @@ char16_t *xvasprintf_status(EFI_STATUS status, const char *format, va_list ap) {
 
 #if SD_BOOT
 /* To provide the actual implementation for these we need to remove the redirection to the builtins. */
+#  undef memchr
 #  undef memcmp
 #  undef memcpy
 #  undef memset
 #else
 /* And for userspace unit testing we need to give them an efi_ prefix. */
+#  define memchr efi_memchr
 #  define memcmp efi_memcmp
 #  define memcpy efi_memcpy
 #  define memset efi_memset
 #endif
 
+_used_ void *memchr(const void *p, int c, size_t n) {
+        if (!p || n == 0)
+                return NULL;
+
+        const uint8_t *q = p;
+        for (size_t i = 0; i < n; i++)
+                if (q[i] == (uint8_t) c)
```

Comment:
hmm, i'd cast this to `(unsigned)` tbh, so that `memchr(p, 1000, n)` will always fail instead of finding a byte `1000 & 0xFF`.

but dunno, i guess this becomes undefined territory

Reply:
Cast what? There is no need for casting thanks to int promotion, use `q[i] == c & UCHAR_MAX` if you want to be pedantic

Reply:
As suggested, `q[i] == (unsigned) c` is better? Otherwise, I guess `memchr("abc", 'a'+0x100, 3) == memchr("abc", 'a', 3)`.

Reply:
> Otherwise, I guess `memchr("abc", 'a'+0x100, 3) == memchr("abc", 'a', 3)`

For better or worse, that's exactly what the spec wants: `c` is to be interpreted as `unsigned char` and compiler explorer agrees too.

In fact, my suggestion to truncate was also wrong. And to be really pedantic, this should really use unsigned char and not uint8_t, as they are technically different.

Reply:
Also, this should probably be asserted in a test case.